File: iosApp/iosApp/ui/TodoAppScreen.swift
//
// Created by Philosopher on 08/08/25.
//


import Foundation
import SwiftUI
import shared



struct TodoAppScreenView: View {
    
    // @StateObject ensures the ViewModel's lifecycle is tied to this view's lifecycle.
    // Use @ObservedObject if the ViewModel is created and managed by a parent and passed in.
    //    @StateObject private var viewModel: MyViewModel = MyViewModel()
    
    @StateObject var viewModelStoreOwner = IOSViewModelStoreOwner()
    @EnvironmentObject var appContainer: ObservableValueWrapper<AppContainer>
    
    var body: some View {
        
        let mainViewModel: MyViewmodel = viewModelStoreOwner.viewModel(
            factory: appContainer.value.todoViewModelFactory
        )
        
        NavigationStack {
            Observing(mainViewModel.uiState) { uiState in // uiState here is the actual value from the flow
                // The `uiState` from `Observing` will be the Kotlin sealed class instance
                // You need to switch on its type. Kotlin/Native generates specific names.
                // For a sealed class `UiState`, the cases might be `UiStateLoading`, `UiStateSuccess`, `UiStateError`.
                // Or, if your KMP library provides helpers, it might be more direct.
                
                // Assuming direct casting or type checking works for the generated Swift enum
                if uiState is UiStateLoading { // Check specific generated type
                    VStack { // Equivalent to Column
                        ProgressView() // Equivalent to CircularProgressIndicator
                        Text("Loading todos...")
                            .padding(.top, 16)
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .center) // .align(Alignment.Center)
                } else if let successState = uiState as? UiStateSuccess<NSArray> { // Assuming List<TodoItem> becomes NSArray
                    // You might need to cast `successState.data` to the correct Swift array type
                    let todos = successState.data as? [Todo] ?? []
                    if todos.isEmpty {
                        Text("No todos found. Add Some!")
                            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .center)
                    } else {
                        TodoListContentView(todos: todos)
                            .frame(maxWidth: .infinity, maxHeight: .infinity) // .fillMaxSize()
                    }
                } else if let errorState = uiState as? UiStateError {
                    ErrorContentView(message: errorState.message)
                        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .center) // Center error content
                } else {
                    // Fallback or initial state if UiState is something else or not yet loaded
                    Text("Unhandled state or initializing...")
                        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .center)
                }
            }
            .navigationTitle("Todos")
        }
    }
}


struct TodoListContentView: View {
    var todos: [Todo]
    var body: some View {
        List {
            ForEach(todos, id: \.id) { todo in
                TodoItemView(todo: todo)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity) // .fillMaxSize()
        .padding(.horizontal, 16) // .padding(horizontal = 16.dp)
    }
}


struct TodoItemView: View {
    let todo: Todo
    
    var body: some View {
        HStack(alignment: .center, spacing: 16) {
            Image(systemName: todo.isCompleted ? "checkmark.square.fill" : "square")
                .font(.title2)
                .foregroundColor(todo.isCompleted ? .accentColor : .secondary)
                .onTapGesture {
                    print("Checkbox tapped for todo: \(todo.title)")
                }
            
            VStack(alignment: .leading, spacing: 4) { // Corresponds to Column
                Text(todo.title)
                    .font(.headline) // Approximate for titleMedium
                    .strikethrough(todo.isCompleted, color: .gray) // For TextDecoration.LineThrough
                    .foregroundColor(todo.isCompleted ? .gray : .primary) // .primary is like onSurface
                
                Text("User ID: \(todo.userId)")
                    .font(.caption) // Approximate for bodySmall
                    .italic()       // For FontStyle.Italic
            }
            .layoutPriority(1) // Helps this VStack take available space, similar to weight(1f)
            
        }
        .padding(16) // Padding for the entire item
        .frame(maxWidth: .infinity, alignment: .leading) // Ensures the HStack tries to fill width, content aligned left
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(Color(UIColor.systemBackground)) // Adapts to light/dark mode
                .shadow(color: .gray.opacity(0.4), radius: 3, x: 0, y: 2) // Basic shadow for card effect
        )
    }
}


struct ErrorContentView: View {
    let message: String
    
    var body: some View {
        VStack(spacing: 16) {
            Text("😔")
                .font(.system(size: 48))
                .foregroundColor(.red)
            
            //            Spacer.frame(height: 1=6)
            
            Text("An error occurred:")
                .font(.headline)
            
            Text(message)
                .font(.body)
                .foregroundColor(.red)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity) // Makes the VStack fill all available space.
        .padding(16)
    }
}

File: composeApp/src/androidMain/kotlin/com/example/kmmtodos/ui/widgets/TodoAppScreen.kt
@file:OptIn(ExperimentalMaterial3Api::class)

package com.example.kmmtodos.ui.widgets

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeContentPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.kmmtodos.LocalAppContainer
import com.example.shared.data.model.Todo
import com.example.shared.ui.MyViewmodel
import com.example.shared.ui.UiState
import androidx.lifecycle.viewmodel.compose.viewModel

@Composable
fun TodoAppScreen(
    // Your ViewModel injection method remains the same.
    viewModel: MyViewmodel = viewModel(
        factory = LocalAppContainer.current.todoViewModelFactory,
    )
) {
    // Corrected to use the new `uiState` property from the ViewModel.
    val uiState by viewModel.uiState.collectAsState()


    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Todo List (Compose)") },
            )
        }
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues),
            contentAlignment = Alignment.Center
        ) {
            when (val currentState = uiState) {
                is UiState.Loading -> {
                    Column(
                        modifier = Modifier.align(Alignment.Center),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        CircularProgressIndicator()
                        Text(
                            "Loading todos...",
                            modifier = Modifier.padding(top = 16.dp)
                        )
                    }
                }

                is UiState.Success -> {
                    val todos = currentState.data
                    if (todos.isEmpty()) {
                        Text(
                            "No todos found. Add Some!",
                            modifier = Modifier.align(Alignment.Center)
                        )
                    } else {
                        // Pass the list to your content Composable.
                        TodoListContent(todos = todos, modifier = Modifier.fillMaxSize())
                    }
                }

                is UiState.Error -> {
                    ErrorContent(message = currentState.message)
                }
            }
        }
    }
}
@Composable
fun TodoListContent(todos: List<Todo>, modifier: Modifier = Modifier) {

    LazyColumn(
        modifier = modifier
            .fillMaxSize()
            .padding(horizontal = 16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        items(
            items = todos,
            key = { todo -> todo.id }

        ) { todo ->

            TodoItem(todo = todo)
//            HorizontalDivider()
        }
    }
}
@Composable
fun TodoItem(todo: Todo, modifier: Modifier = Modifier) {

    Card(
        modifier = modifier
            .fillMaxWidth(),
        shape = RoundedCornerShape(8.dp)


    ) {

        Row(
            modifier = Modifier
                .padding(all = 16.dp)
                .fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Checkbox(
                checked = todo.isCompleted,
                onCheckedChange = {}
            )
            Spacer(modifier = Modifier.width(16.dp))
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = todo.title,
                    style = MaterialTheme.typography.titleMedium,
                    textDecoration = if (todo.isCompleted) TextDecoration.LineThrough else null,
                    color = if (todo.isCompleted) Color.Gray else MaterialTheme.colorScheme.onSurface

                )
                Spacer(modifier = Modifier.height(4.dp))

                Text(
                    text = "User ID: ${todo.userId}",
                    style = MaterialTheme.typography.bodySmall,
                    fontStyle = FontStyle.Italic
                )
            }
        }
    }
}

@Composable
fun ErrorContent(message: String, modifier: Modifier = Modifier) {

    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
//        Log.d("TAG", "ErrorContent: $message")

        Text(
            text = "😔", // Sad emoji
            fontSize = 48.sp,
            color = MaterialTheme.colorScheme.error
        )
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "An error occurred:",
            style = MaterialTheme.typography.titleMedium
        )
        Text(
            text = message,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.error
        )
    }
}

